lgc/test/Transforms/CpsLowering/cps-unify-exits.lgc

```diff
@@ -0,0 +1,202 @@
+; NOTE: Assertions have been autogenerated by tool/update_llpc_test_checks.py UTC_ARGS: --tool lgc --function-signature
+; RUN: lgc -mcpu=gfx1030 -o - -passes='require<lgc-pipeline-state>,lgc-patch-entry-point-mutate' %s | FileCheck --check-prefixes=CHECK %s
+
+declare void @lgc.cps.jump(...) noreturn
+
+define void @unify_jumps({i32} %state, i32 %arg, ptr %table) !lgc.cps !0 {
+entry:
+  %v = extractvalue {i32} %state, 0
+  %cond = icmp ult i32 %v, 3
+  br i1 %cond, label %then, label %else
+
+then:
+  %table.0 = getelementptr i32, ptr %table, i32 0
+  %cr.then = load i32, ptr %table.0
+  %then.arg = add i32 %arg, 1
+  %v.then = mul i32 %v, 2
+  %state.then = insertvalue {i32} poison, i32 %v.then, 0
+  call void (...) @lgc.cps.jump(i32 %cr.then, i32 2, {i32} %state.then, i32 %then.arg)
+  unreachable
+
+else:
+  %table.1 = getelementptr i32, ptr %table, i32 1
+  %cr.else = load i32, ptr %table.1
+  %else.arg = uitofp i32 %arg to float
+  call void (...) @lgc.cps.jump(i32 %cr.else, i32 2, {} poison, float %else.arg, i32 5)
+  unreachable
+}
+
+define void @unify_jump_ret({i32} %state, i32 %arg, ptr %table) !lgc.cps !0 {
+entry:
+  %v = extractvalue {i32} %state, 0
+  %cond = icmp ult i32 %v, 3
+  br i1 %cond, label %then, label %else
+
+then:
+  %table.0 = getelementptr i32, ptr %table, i32 0
+  %cr.then = load i32, ptr %table.0
+  %then.arg = add i32 %arg, 1
+  %v.then = mul i32 %v, 2
+  %state.then = insertvalue {i32} poison, i32 %v.then, 0
+  call void (...) @lgc.cps.jump(i32 %cr.then, i32 2, {i32} %state.then, i32 %then.arg)
+  unreachable
+
+else:
+  ret void
+}
+
+!0 = !{i32 1} ; level 1
+; CHECK-LABEL: define {{[^@]+}}@unify_jumps.1
+; CHECK-SAME: (i32 inreg [[GLOBALTABLE:%.*]], i32 inreg [[PERSHADERTABLE:%.*]], ptr addrspace(4) inreg [[NUMWORKGROUPSPTR:%.*]], i32 inreg [[TMP0:%.*]], i32 inreg [[TMP1:%.*]], i32 inreg [[TMP2:%.*]], i32 inreg [[TMP3:%.*]], i32 inreg [[TMP4:%.*]], i32 inreg [[TMP5:%.*]], i32 inreg [[TMP6:%.*]], i32 inreg [[TMP7:%.*]], i32 inreg [[TMP8:%.*]], i32 inreg [[TMP9:%.*]], i32 inreg [[TMP10:%.*]], i32 inreg [[SPILLTABLE:%.*]], i32 [[VCR:%.*]], ptr addrspace(5) [[VSP:%.*]], i32 [[ARG:%.*]], ptr [[TABLE:%.*]]) #[[ATTR1:[0-9]+]] align 64 !lgc.cps !2 {
```

Comment:
The CI test seems failed because the metadata number changed from `!2` to `!3` on the CI machine. And I even cannot reproduce the failure locally... I have experienced such issue when I ported the change from internal, at that time the number changed from `!0` to `!2`. So anybody know a good way to handle such case. It is annoying that the number keep changing.

Reply:
The more strange thing is the test only fails for clang, but passed for gcc...

Reply:
You can try to capture the number with ![[name:[0-9]+]] and then check in the same test run that the captured node is equivalent to some value with

CHECK: ![[name]] = !...

This requires to run different tests with other names and a bit of tinkering.

Reply:
I think in this particular case, we are not interested in the metadata contents, it's rather a UTC-generated test that includes MD node ids, which seem to change for some reason.

We need to understand why they are changing. If it cannot be avoid, we could add a UTC flag to ignore MD node ids, but clang and gcc already disagree this sounds more like a bug somewhere.

Reply:
I only see 3 metadata entries, so I wonder where another one is coming from.
```llvm
!0 = !{!""}
!1 = !{!"\82\B0amdpal.pipelines\91\83\AA.registers\83\CD.@\CE\10\00\00\00\CD.B\CE\10\00\00\06\CD.O\CE\10\00\00\02\B0.spill_threshold\CE\FF\FF\FF\FF\B0.user_data_limit\00\AEamdpal.version\92\02\06"}
!2 = !{i32 1}
```

We really want deterministic compilation (and an internal CI job checks that at least the outcoming ELF is consistent) :)

Reply:
The weird extra metadata in clang build comes from the `m_options` in PipelineState (https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/include/lgc/state/PipelineState.h#L555). The sizeof(Options) is 128 bytes, actually the last field is at offset 122, so there are 6 tail padding bytes. The brace initialization sounds like does not initialize the padding 6bytes in clang build. But it was read inside `setNamedMetadataToArrayOfInt32`. That's where `i32 -16777216, i32 -1` comes from.

Reply:
Yes, I think the solution is to add explicit padding everywhere in these structures.

Reply:
I think maintaining explicitly padded struct is annoying. I have just made a change #2584 following the way ShaderOptions being zero initialized.